### DISCUSSION TITLE
[RFC] Penggunaan Temporal.io untuk orchestration

### DISCUSSION THREAD

**Discussion by aldy505** [TokoBapak/.github#17] [Backend Team Posts]
Dikarenakan ini project e-commerce microservice, dimana sudah dipastikan bahwa sistem akan berupa event-driven architecture, maka tidak akan jauh dari saga pattern. Walaupun sudah ada konsensus awal [untuk menggunakan gRPC](https://github.com/TokoBapak/.github/blob/3d81b5c2eb37a3ccc12a04a6bade8c0f8710f7dc/NEW_BACKEND_SERVICE.md) karena ini merupakan project bersifat hype-driven, dengan jumlah kontributor yang sedikit nampaknya akan sulit apabila kita memaksakan mengurus orchestration beserta segala permasalahannya.

RFC ini bertujuan untuk mengajukan penggunaan [Temporal](https://temporal.io/) untuk orchestration. Perlu dicatat bahwa Temporal ini hanya service orchestration, untuk presentation layer ke customer tetap menggunakan gRPC/HTTP layer terpisah dari Temporal.

Pros:
* Mempermudah communication antar service, tidak perlu mock suatu service kalau service belum jadi.
* Beberapa problem dalam saga pattern sudah di handle seperti retry, timeout, schedule

Cons:
* Language support sedikit: Go, Java, PHP, Python, Typescript, .NET, Rust (masih alpha, [masih banyak issue](https://github.com/temporalio/sdk-core/issues?q=is%3Aopen+is%3Aissue+label%3A%22Rust+SDK%22)). Plain gRPC yang awalnya mau kita pakai punya lebih banyak language support 
* Requires additional infra being set up (see [deployment guide](https://docs.temporal.io/self-hosted-guide/setup))

/cc @TokoBapak/backend @TokoBapak/infrastructure

**Comment by oinpentuls** [comment]
> Perlu dicatat bahwa Temporal ini hanya service orchestration, untuk presentation layer ke customer tetap menggunakan gRPC/HTTP layer terpisah dari Temporal.

Dari diagram ini, temporal bakal ada di mana pak ?

![image](https://github.com/TokoBapak/.github/assets/39396616/18687380-1bc8-4b29-b086-83ae9b19e20d)

**Reply by aldy505** [reply]
Bersebelahan dengan message bus, karena kan dia workflow control server.

**Reply by oinpentuls** [reply]
okay
